examples/ode-integrator.dx

```diff
@@ -1,22 +1,41 @@
+'## Euler Integrator
+Using only generic vector operations.
 
-EState : Type -> Type = \d. d=>Real
-Time :Type = Real
-Dynamics : Type -> Type = \d. EState d -> EState d
+Time : Type = Real
 
+def eulerStep (dict: VSpace a) ?=>
+              (zref: Ref h a) ->
+              (dynamics: a -> Time -> a) ->
+              (dt: Time) ->
+              (t: Time)
+              : {State h} Unit =
+    z = get zref
+    zref := z + dt .* dynamics z t
 
-eulerStep : Ref h (EState d) -> Dynamics d -> Time -> {State h} Unit =
-  \sRef dyn dt.
-    s = get sRef
-    sRef := s + dt .* dyn s
+def euler (dict: VSpace a) ?=> 
+          (dynamics: a -> Time -> a) ->
+          (z0: a) ->
+          (t0: Time) ->
+          (t1: Time) ->
+          (dt: Time)
+          : a =
+     n = floor ((t1 - t0) / dt)
+     snd $ withState z0 \zref.
+       for i:(Fin n).
+         t = t0 + (i2r (ordinal i)) * dt
```

Comment:
Wouldn't it be nicer to just `eulerStep` over a `linspace` of `dt`s instead of doing the math here?

Reply:
Yes, you're right!